compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala

```diff
@@ -61,82 +73,42 @@ class ShortcutImplicits extends MiniPhase with IdentityDenotTransformer { thisPh
-    *  to be only targets of monomorphic calls because they are effectively
-    *  final and don't override anything.
-    */
-  private def shouldBeSpecialized(sym: Symbol)(implicit ctx: Context) =
-    sym.is(Method, butNot = Accessor) &&
-    defn.isImplicitFunctionType(sym.info.finalResultType) &&
-    defn.functionArity(sym.info.finalResultType) > 0 &&
-    !sym.isAnonymousFunction &&
-    (specializeMonoTargets || !sym.isEffectivelyFinal || sym.allOverriddenSymbols.nonEmpty)
-
-  /** @pre    The type's final result type is an implicit function type `implicit Ts => R`.
-    *  @return The type of the `apply` member of `implicit Ts => R`.
-    */
-  private def directInfo(info: Type)(implicit ctx: Context): Type = info match {
-    case info: PolyType   => info.derivedLambdaType(resType = directInfo(info.resultType))
-    case info: MethodType => info.derivedLambdaType(resType = directInfo(info.resultType))
-    case info: ExprType   => directInfo(info.resultType)
-    case info             => info.member(nme.apply).info
-  }
-
-  /** A new `m$direct` method to accompany the given method `m` */
-  private def newDirectMethod(sym: Symbol)(implicit ctx: Context): Symbol = {
-    val direct = sym.copy(
-      name = DirectMethodName(sym.name.asTermName).asInstanceOf[sym.ThisName],
-      flags = sym.flags | Synthetic,
-      info = directInfo(sym.info))
-    if (direct.allOverriddenSymbols.isEmpty) direct.resetFlag(Override)
-    direct
-  }
 
   /** The direct method `m$direct` that accompanies the given method `m`.
     *  Create one if it does not exist already.
     */
   private def directMethod(sym: Symbol)(implicit ctx: Context): Symbol =
-    if (sym.owner.isClass) {
-      val direct = sym.owner.info.member(DirectMethodName(sym.name.asTermName))
-        .suchThat(_.info matches directInfo(sym.info)).symbol
-      if (direct.maybeOwner == sym.owner) direct
-      else newDirectMethod(sym).enteredAfter(thisPhase)
-    }
-    else directMeth.getOrElseUpdate(sym, newDirectMethod(sym))
+    if (sym.owner.isClass) shortcutMethod(sym, thisPhase)
+    else directMeth.getOrElseUpdate(sym, newShortcutMethod(sym))
 
   /** Transform `qual.apply` occurrences according to rewrite rule (2) above */
   override def transformSelect(tree: Select)(implicit ctx: Context) =
     if (tree.name == nme.apply &&
         defn.isImplicitFunctionType(tree.qualifier.tpe.widen) &&
-        shouldBeSpecialized(tree.qualifier.symbol)) {
+        needsImplicitShortcut(tree.qualifier.symbol)) {
       def directQual(tree: Tree): Tree = tree match {
         case Apply(fn, args)     => cpy.Apply(tree)(directQual(fn), args)
         case TypeApply(fn, args) => cpy.TypeApply(tree)(directQual(fn), args)
         case Block(stats, expr)  => cpy.Block(tree)(stats, directQual(expr))
         case tree: RefTree =>
+          def rewire(tp: Type, sym: Symbol) = tp match {
+            case tp: NamedType => tp.prefix.select(sym)
+            case _ => sym.termRef
```

Comment:
Could you explain when the last case is reachable?

Reply:
I'll try to eliminate the last case. It's better to fail fast if we misunderstand something.